src/java.base/share/classes/java/lang/template/StringTemplate.java

```diff
@@ -0,0 +1,368 @@
+ * expression with the standard {@link StringTemplate#STR} processor.
+ * {@snippet :
+ * int x = 10;
+ * int y = 20;
+ * String result = STR."\{x} + \{y} = \{x + y}";
+ * }
+ * {@code result} will be equivalent to <code>"10 + 20 = 30"</code>.
+ * <p>
+ * The {@link StringTemplate#process(ValidatingProcessor)} method supplies an
+ * alternative to using string template expressions.
+ * {@snippet :
+ * String result = RAW."\{x} + \{y} = \{x + y}".process(STR);
+ * }
+ * In addition to string template expressions, the factory methods
+ * {@link StringTemplate#of(String)} and {@link StringTemplate#of(List, List)}
+ * can be used to construct {@link StringTemplate StringTemplates}.
+ * <p>
+ * The {@link StringTemplate#interpolate()} method provides a simple way to produce a
+ * string interpolation of the {@link StringTemplate}.
+ * <p>
+ * {@linkplain ValidatingProcessor Template processors} typically use the following code
+ * pattern to perform composition:
+ * {@snippet :
+ * List<String> fragments = st.fragments();
+ * List<Object> values = st.values();
+ * // check or manipulate the fragments and/or values
+ * ...
+ * String result = TemplateRuntime.interpolate(fragments, values);;
+ * }
+ *
+ * @implSpec An instance of {@link StringTemplate} is immutatble. Also, the
+ * fragment list size must be one more than the values list size.
+ *
+ * @see ValidatingProcessor
+ * @see TemplateProcessor
+ * @see StringProcessor
+ * @see java.util.FormatProcessor
+ *
+ * @since 20
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+public interface StringTemplate {
+    /**
+     * Returns an immutable list of string fragments consisting of the string
+     * of characters preceeding each of the embedded expressions plus the
+     * string of characters following the last embedded expression. In the
+     * example: {@snippet :
+     * StringTemplate st = RAW."The student \{student} is in \{teacher}'s class room.";
+     * List<String> fragments = st.fragments(); // @highlight substring="fragments()"
+     * }
+     * <code>fragments</code> will be equivalent to
+     * <code>List.of("The student ", " is in ", "'s class room.")</code>
+     *
+     * @return list of string fragments
+     *
+     * @implSpec The list returned is immutable.
+     */
+    List<String> fragments();
+
+    /**
+     * Returns an immutable list of embedded expression results. In the example:
+     * {@snippet :
+     * StringTemplate st = RAW."\{x} + \{y} = \{x + y}";
+     * List<Object> values = st.values(); // @highlight substring="values()"
+     * }
+     * <code>values</code> will be equivalent to <code>List.of(x, y, x + y)</code>
+     *
+     * @return list of expression values
+     *
+     * @implSpec The list returned is immutable.
+     */
+    List<Object> values();
+
+    /**
+     * {@return the interpolation of the StringTemplate}
+     */
+    default String interpolate() {
```

Comment:
I wonder if all the default methods should not be better as static methods given that they are not the important part of the API but more side information that may be handy

Reply:
Actually instance interpolate() is the most important method. Each synthetic StringTemplate gets a specialized interpolate providing performance equivalent to string concat. And, a good percentage  of processors will work with the result of interpolate to produce result. Ex. `StringProcessor STR = st -> st.interpolate();` and`TemplateProcessor<JSONObject> JSON = st -> new JSONObject(st.interpolate());`

Reply:
Will add